### PR TITLE
Added match API to lookup items using partial key

### PR DESCRIPTION
**Pull request by matyasmarkovics**
This change allows caching an item with a compound key and look up a list of items matching only some part of the key.

Changes also include bringing the makefile up-to-date.

### CODE REVIEW

File: src/erl_cache_server.erl [hunk 1]
```diff
@@ -69,24 +70,55 @@ start_link(Name) ->
...
 
+send_stat(Name, Now, #cache_entry{validity=Validity}) when Now < Validity ->
+    gen_server:cast(Name, {increase_stat, hit});
+send_stat(Name, Now, #cache_entry{evict=Evict}) when Now < Evict ->
+    gen_server:cast(Name, {increase_stat, overdue});
+send_stat(_Name, _Now,  _) ->
+    ok.
+
+get_valid_value(Name, WaitForRefresh, Now, Entry) ->
+    case Entry of
```
Comment (motobob) [inline src/erl_cache_server.erl]: stylish: case on function param should be function clause.

File: src/erl_cache_server.erl [hunk 2]
```diff
@@ -69,24 +70,48 @@ start_link(Name) ->
...
+    {ok, erl_cache:value()} | {error, not_found}.
+match(Name, Key, WaitForRefresh) ->
+    Now = now_ms(),
+    case ets:match_object(get_table_name(Name), #cache_entry{key = Key, _ = '_'}) of
+        [] ->
+            gen_server:cast(Name, {increase_stat, miss}),
+            {error, not_found};
+        Matches ->
+            lists:foreach(fun(E) -> send_stat(Name, Now, E) end, Matches),
+            IsValid = fun(E) -> get_valid_value(Name, WaitForRefresh, Now, E) end,
```
Comment (kkuzmin) [inline src/erl_cache_server.erl]: I think in order to keep consistent behavior and perform less list operations `IsValid` should be:
```
fun(E) -> 
    send_stat(Name, Now, E),
    get_valid_value(Name, WaitForRefresh, Now, E)
end,
```

File: src/erl_cache_server.erl [hunk 3]
```diff
@@ -69,24 +70,48 @@ start_link(Name) ->
...
+get_valid_value(_Name, _IsSync, _Now, #cache_entry{value=_ExpiredValue}) ->
+    false.
+
+-spec match(erl_cache:name(), erl_cache:key(), erl_cache:wait_for_refresh()) ->
+    {ok, erl_cache:value()} | {error, not_found}.
+match(Name, Key, WaitForRefresh) ->
+    Now = now_ms(),
+    case ets:match_object(get_table_name(Name), #cache_entry{key = Key, _ = '_'}) of
+        [] ->
+            gen_server:cast(Name, {increase_stat, miss}),
```
Comment (kkuzmin) [inline src/erl_cache_server.erl]: It feels like `{increase_stat, miss}` should also become a part of `send_stat()` function interface. But I can't think of a good way to make it so. Maybe,
```
send_stat(Name, _Now, undefined/miss) ->
    gen_server:cast(Name, {increase_stat, miss});
```

File: src/erl_cache_server.erl [hunk 4]
```diff
@@ -241,21 +266,21 @@ purge_cache( Name, TableName, Now ) ->
...
 %% @private
 -spec do_refresh(erl_cache:name(), #cache_entry{}, erl_cache:wait_for_refresh()) ->
     erl_cache:value().
-do_refresh(Name, #cache_entry{key=Key, validity_delta=ValidityDelta, evict_delta=EvictDelta,
-                              refresh_callback=Callback, is_error_callback=IsErrorCb}=Entry,
-           WaitForRefresh) ->
+do_refresh(Name, #cache_entry{} = Entry, WaitForRefresh) ->
+    #cache_entry{key=Key, validity_delta=ValidityDelta, evict_delta=EvictDelta,
+                 refresh_callback=Callback, is_error_callback=IsErrorCb} = Entry,
+    ?DEBUG("Refreshing overdue key ~p", [Key]),
```
Comment (kkuzmin) [inline src/erl_cache_server.erl]: I think `Name` output will also help with debugging.

**Review by kkuzmin — CHANGES_REQUESTED** [review]

**Review by motobob — APPROVED** [review]
agree with @kkuzmin comments.
otherwise +1

File: src/erl_cache_server.erl (hunk 3 above)
Comment (matyasmarkovics) [reply src/erl_cache_server.erl]: Please check what I've done to preserve the function signature.

**Review by kkuzmin — APPROVED** [review]

**Review by motobob — APPROVED** [review]

**Pull request merged**